manager/controllers/default/security/login.class.php

```diff
@@ -460,7 +556,95 @@ public function handleForgotLogin() {
+            $hash => $user->get('username')
+        ], [
+            'ttl' => $this->modx->getOption('passwordless_expiration')
+        ]);
+    }
+
+    /**
+     * Handles the action when a user requests a magic login link.
+     *
+     * @return void
+     * @throws Exception
+     */
+    public function handlePasswordlessLoginRequest() {
+        $this->modx->log(modX::LOG_LEVEL_ERROR, "Handling passwordless login");
+
+        $c = $this->modx->newQuery('modUser');
+        $c->select(['modUser.*', 'Profile.email', 'Profile.fullname']);
+        $c->innerJoin('modUserProfile', 'Profile');
+        $c->where([
+            'Profile.email:=' => $this->scriptProperties['passwordless_login_email'],
+        ]);
+
+        /** @var modUser $user */
+        $user = $this->modx->getObject('modUser', $c);
+
+        if ($user) {
+            $this->modx->log(modX::LOG_LEVEL_DEBUG, "Sending out magic link for user " . $user->get('username'));
+            // we currently just use the forgot password hash
+
+            $activationHash = md5(uniqid(md5($user->get('email') . '/' . $user->get('id')), true));
+
+            $this->setActivationHash($activationHash, $user);
+
+            // Send activation email
+            $message = $this->modx->lexicon('login_magiclink_email');
+            $placeholders = array_merge($this->modx->config, $user->toArray());
+            $placeholders['hash'] = $activationHash;
+            $placeholders['expiration'] = date('D, d M Y H:i',time() + $this->modx->getOption('passwordless_expiration'));
+
+            // Store previous placeholders
+            $ph = $this->modx->placeholders;
+            // now set those useful for modParser
+            $this->modx->setPlaceholders($placeholders);
+            $this->modx->getParser()->processElementTags('', $message, true, false, '[[', ']]', [], 10);
+            $this->modx->getParser()->processElementTags('', $message, true, true, '[[', ']]', [], 10);
+            // Then restore previous placeholders to prevent any breakage
+            $this->modx->placeholders = $ph;
+
+            $this->modx->smarty->assign('config', $this->modx->config);
+            $this->modx->smarty->assign('content', $message);
+            $message = $this->modx->smarty->fetch('email/default.tpl');
+            $sent = $user->sendEmail($message, [
+                'from' => $this->modx->getOption('emailsender'),
+                'fromName' => $this->modx->getOption('site_name'),
+                'sender' => $this->modx->getOption('emailsender'),
+                'subject' => $this->modx->lexicon('login_magiclink_subject'),
+                'html' => true,
+            ]);
+            if (!$sent) {
+                $this->setPlaceholder('error_message', $this->modx->lexicon('error_sending_magiclink_to'));
+            } else {
+                $this->setPlaceholder('success_message', $this->modx->lexicon('login_user_err_nf_email'));
+            }
+        } else {
```

Comment:
This `else` is exposing an enumeration vulnerability allowing valid (registered) emails to be bruteforced. 

To avoid that vulnerability, the response has to be identical for users that exist and users that don't. Ideally it also needs to account for the processing that happens when a user exists (parsing a message, sending an email), ensuring that timing can't be used to distinguish between a successful/failed attempt.

Reply:
> Ideally it also needs to account for the processing that happens when a user exists (parsing a message, sending an email), ensuring that timing can't be used to distinguish between a successful/failed attempt.

I guess the same should apply for regular logins and password resets?

Reply:
I think it's not necessary for logins as there's either a failure or success, plus user-level brute force protection, and IIRC we addressed the difference between "user does not exist" and "password for user is incorrect" in 2.6 or so. 

For password resets, yes, same protections need to apply there, but IIRC that was already addressed for 3.x when the password reset was reworked.

Reply:
Both cases now return the same message to the login frontend. 
Regarding the timing, how would you expect that to be implemented? Timing difference is minimal and processing varies a lot here. It might be possible to sleep the processing so that both e.g. take the same 1s to respond. But is that what we want here? Or just add random 200-500ms to the process? Still it depends on the connectivity when sending out the mail... it might take longer to send out an email (shouldnt, but...)

Reply:
@wuuti Here's an article that deals with the varying ways to prevent timing attacks: https://blog.ircmaxell.com/2014/11/its-all-about-time.html#An-Actual-Delay-That-Works